proposals/p2550.md

```diff
@@ -0,0 +1,218 @@
+    `Foo` is not allowed to be the name of the current package.
+-   `import library "Bar";` to import library `"Bar"` of the current package.
+    This introduces the names from that library at file scope. `"Bar"` is not
+    allowed to be the name of the current library.
+
+There is no way to import libraries from the main package from any other
+package.
+
+It is permitted for a Carbon program to contain a source file that has no
+`package` declaration and does not contain a `Main` function -- or even to
+contain multiple such source files. Such a source file cannot implement any
+importable functionality, as there is no corresponding `api` file, but may still
+be useful to support Carbon features that have not yet been designed, such as:
+
+-   Global registration mechanisms.
+-   Some way of defining functions with well-known symbols, analogous to
+    `extern "C"` declarations in C++.
+
+## Details
+
+See design changes.
+
+## Rationale
+
+-   [Community and culture](/docs/project/goals.md#community-and-culture)
+    -   This change allows small complete programs and source files to be
+        discussed in Carbon forums without the overhead of `package`
+        declarations. Cumbersome top-level syntax hampers technical discussion.
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   Current test cases for `explorer` are made unnecessarily verbose by the
+        mandatory inclusion of a `package` declaration and a `Main` function.
+        This proposal makes these components optional.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   Renaming a package requires changing only the first line of each source
+        file, making this operation somewhat easier.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Small test cases and `"hello world"` programs have less boilerplate.
+    -   Reduction of verbosity that is not contributing to increased
+        understanding of the program.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   This proposal explicitly acknowledges and permits the entry point for a
+        program that includes Carbon code to be written in a different language.
+
+## Alternatives considered
+
+### Give the main package an identifier name
+
+We could give the main package an identifier name, such as `Main`, and require
+that name to be used in `package` declarations:
+
+```
+package Main impl;
+import SimpleIO;
+
+fn Main() {
+  SimpleIO.Print("Hello, Carbon!");
+}
+```
+
+This would keep the Carbon language more consistent, and remove a special case.
+However, it would also increase verbosity in the simplest of programs, where the
+added verbosity has the most cost.
+
+### Distinguish file scope from package scope
```

Comment:
Can you please address how you address name lookup ambiguity that occurs because of not having a way to refer to the package scope? i.e., how do I make the following code work:

```
fn Foo();

class Bar {
  fn Foo();
  fn Wiz() {
    Foo();  // How do I get this to refer to mypackage.Foo()?
  }
}
```

Or:

```
import Foo;

class Bar {
  fn Foo();
  fn Wiz() {
    Foo.Bar();  // Again, I want to call Foo.Bar(), but name lookup probably runs into mypackage.Bar.Foo()
  }
}
```

Are package level entities disallowed from sharing identifiers with any other entity within the package?

Reply:
Under #1136, this is done by using the `package` keyword as a symbolic name for the current package scope / file scope. I felt that adding that to this proposal would be too much scope creep. Added some words about that to this section.